73 configuring assignments for self reviews

laurenbrissette opened the pull request:
- add "self_review_settings" where each "assignment" row has the id of it's "self_review_settings" entry
- "self_review_settings" have fields: enabled, deadline offset (from assignment due date), allow_early, and class_id
- auto create a self review rubric when the assignment is created
- add control for student to finish their assignment early and get the self review (if allow_early is enabled)
-->  this creates a new assignment due date exception for themselves, with the number of hours extended being a negative number to represent their deadline has been moved up.  no late tokens are used
-->  an insert in the "assignment_due_date_exceptions" table causes the sql function "auto_assign_self_reviews" to run, which creates review assignments using the self review rubric for any students that should have a self review assignment based on the current time in relation to their assignment deadline, but do not yet have one
- add cron process that runs every minute and assigns self reviews to profiles with an assignment deadline that just passed 

future support needed:
- ui for students to view their review assignments

laurenbrissette committed:
rls for self eval settings + add fields to assignment form

laurenbrissette committed:
Update page.tsx

laurenbrissette committed:
error handling in form + cleanup

laurenbrissette committed:
enhanced review setting setup

- all assignments must have a self review setting, even if it is disabled
- add foreign keys for self review settings
- seed file updated to create a self review setting with the assignment

laurenbrissette committed:
draft automate assigning self reviews

laurenbrissette committed:
add more comments to func + auto trigger function after student finalizes

laurenbrissette committed:
group + late day rules

- no late days after finishing assignment early
- identify assignment due dates by student profiles OR group ids

laurenbrissette committed:
separate out logic for individual assignment + profile combination

jon-bell commented on app/course/[course_id]/assignments/[assignment_id]/finalizeSubmissionEarly.tsx:
```diff
@@ -0,0 +1,99 @@
+          assignment_group_id: group_id,
+          group_id: memberGroup.data[0].assignment_group_id,
+          creator_id: private_profile_id,
+          hours: hoursBetween(assignment.due_date, new Date().toDateString()), // difference between assignment due date and now, should be a negative number
```
FYI, once I update the due dates to be specified in a timezone, I do not think that this calculation will work. However, it should work using the date-fns/tz helpers.

jon-bell commented on supabase/migrations/20250523191212_assignment-eval-settings.sql:
```diff
@@ -0,0 +1,245 @@
+AS PERMISSIVE 
+FOR INSERT
+WITH CHECK (
+   "hours" < 0 
```
I can't easily tell how all of the policies interact, but: is there something else that restricts students to only be able to insert negative extensions for themselves?

jon-bell replied:
Sorry, what I meant was: I think that the check should also enforce `authorizeforprofile(student_id) or authorizeforassignmentgroup(assignment_group_id)`, to prevent a student from entering a negative extension for another student (unless there is something else that is enforcing this, but I didn't think so)

jon-bell commented on supabase/migrations/20250523191212_assignment-eval-settings.sql:
```diff
@@ -0,0 +1,245 @@
+end;$function$
+;
+
+-- TEMPORARY: REFINE!!!!!!!
```
I think that the simplest security model for this RLS (and the auto_assign_self_reviews function) is:
- Only instructors can directly insert review_assignments
- auto_assign_self_reviews can not be invoked by anyone other than postgres, gets set as SECURITY DEFINER

This way, when the auto_assign_self_reviews function is invoked by a trigger (using logic that we control), it can create review_assignments.

laurenbrissette committed:
scheduling cron job for regular review assignment

scan assignments when NOW() is approx. between the assignment deadline and the late deadline

laurenbrissette committed:
use all assignments for class to catch exceptions

can adjust to be more specific as needed if uses too many resources

use hours function for timezones

laurenbrissette committed:
merge main

laurenbrissette committed:
Merge branch 'staging' of https://github.com/pawtograder/platform into 73-configuring-assignments-for-self-reviews

laurenbrissette committed:
fix db types

laurenbrissette committed:
timezones in deadline calculations

laurenbrissette committed:
add minutes to extensions + better group handling

laurenbrissette committed:
Update 20250523191212_assignment-eval-settings.sql

laurenbrissette committed:
debug race condition in single review creation

laurenbrissette committed:
fix group id undefined

laurenbrissette committed:
give default class time zone

laurenbrissette committed:
cleanup

laurenbrissette committed:
Update 20250523191212_assignment-eval-settings.sql

laurenbrissette commented on supabase/migrations/20250523191212_assignment-eval-settings.sql:
```diff
@@ -0,0 +1,275 @@
+    profile_record public.profiles;
+BEGIN
+    FOR assignment_record IN (
+        SELECT * FROM assignments -- non archived assignment with passed deadline
```
Right now looks at all assignments across classes that haven't been archived, could be narrowed down more if needed to avoid unnecessary checks

jon-bell committed:
Merge branch 'staging' into 73-configuring-assignments-for-self-reviews

jon-bell committed:
Integrate self reviewing into UI. Creates a new hook to store which review we are working on in one place

jon-bell committed:
lint

jon-bell reviewed (dismissed):
Really nice! I merged in recent changes by @Ari-S-123 that put the reviewing UI in a better place to support self-reviews, and also tweaked that work a bit so that there is now a single react hook that can be used anywhere in the submission interface to find out which rubric and review are being worked on. It appears to work from end-to-end now to 1) configure self-review on an assignment, 2) submit self-review as a student!

Here are some notes of issues that I came across that are still open:
- I think that I would prefer the longer "assignment_self_review_settings" to  "self_review_settings", as we might later on have some other forms of self review
- The edit assignment form does not correclty populate with the existing self-review settings
- We need to more clearly show self review information to students: will a self-review be required, and if so, by when? This should be on the assignment overview page, and also when viewing their submission (e.g. in the new `SelfReviewDueDateInformation` 

There is also some other cleanup around the self-review workflow and grading interface to finish the feature, but let's plan those for a separate PR after getting these few things resolved and merged!

jon-bell commented:
Also sorry I clicked "approve", could you please still make the changes I listed? :)

laurenbrissette committed:
rename self review setting table + populate edit assignment form w/ eval

laurenbrissette committed:
add deadlines for students throughour

jon-bell committed:
Continue to refactor hooks to support multiple reviews more seamlessly

jon-bell committed:
Add missing migration data

jon-bell committed:
Merge branch '73-configuring-assignments-for-self-reviews' of github.com:pawtograder/platform into 73-configuring-assignments-for-self-reviews

jon-bell committed:
lint :(

jon-bell reviewed (approved):
LGTM
